crates/red_knot_python_semantic/src/node_key.rs

```diff
@@ -1,24 +1,18 @@
-use ruff_python_ast::{AnyNodeRef, NodeKind};
-use ruff_text_size::{Ranged, TextRange};
+use ruff_python_ast::AnyNodeRef;
 
 /// Compact key for a node for use in a hash map.
 ///
-/// Compares two nodes by their kind and text range.
+/// Stores the memory address of the node, because using the range and node kind
+/// is not enough to uniquely identify them in ASTs resulting from invalid syntax.
 #[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
-pub(super) struct NodeKey {
-    kind: NodeKind,
-    range: TextRange,
-}
+pub(super) struct NodeKey(usize);
```

Comment:
Thought about renaming it to `NodeAddress`, but that lead to too many downstream renamings and a large diff. Let me know if someone is unhappy with keeping `NodeKey`.